p2panda-rs/src/encoder.rs

```diff
@@ -0,0 +1,287 @@
+use std::convert::{TryFrom, TryInto};
+
+use arrayvec::ArrayVec;
+use bamboo_rs_core::entry::MAX_ENTRY_SIZE;
+use bamboo_rs_core::{Entry as BambooEntry, Signature as BambooSignature, YamfHash};
+use ed25519_dalek::PublicKey;
+
+use crate::atomic::error::EntrySignedError;
+use crate::atomic::Blake2BArrayVec;
+use crate::atomic::{Entry, EntrySigned, Hash, LogId, Message, MessageEncoded, SeqNum};
+use crate::key_pair::KeyPair;
+
+/// Takes an [`EntrySigned`] and a [`MessageEncoded`], validates the encoded message against the
+/// entry payload hash, returns the decoded message in the form of a [`Message`] if valid otherwise
+/// throws an error.
+pub fn safe_decode_message(
+    entry_encoded: &EntrySigned,
+    message_encoded: &MessageEncoded,
+) -> Result<Message, EntrySignedError> {
+    // Convert to Entry from bamboo_rs_core first
+    let entry: BambooEntry<ArrayVec<[u8; 64]>, ArrayVec<[u8; 64]>> = entry_encoded.into();
+
+    // Message hash must match if it doesn't return an error
+    let yamf_hash: YamfHash<Blake2BArrayVec> = (&message_encoded.hash()).to_owned().into();
+    if yamf_hash != entry.payload_hash {
+        return Err(EntrySignedError::MessageHashMismatch);
+    }
+
+    Ok(Message::from(message_encoded))
+}
```

Comment:
We can remove the decoding part and name this method `validate_message_entry_match`

Reply:
Even better: Move this into `EntrySigned` and rename method to `validate_message`